Libplanet/Blockchain/Renderers/AtomicActionRenderer.cs

```diff
@@ -0,0 +1,179 @@
+#nullable enable
+using System;
+using System.Collections.Generic;
+using Libplanet.Action;
+using Libplanet.Blocks;
+using Libplanet.Tx;
+
+namespace Libplanet.Blockchain.Renderers
+{
+    /// <summary>
+    /// A middleware to make action render events to satisfy transactions' atomicity.
+    /// <para>Decorates an <see cref="IActionRenderer{T}"/> instance and filters out render events
+    /// made by unsuccessful transactions (i.e., transactions with one or more exception-throwing
+    /// actions).</para>
+    /// </summary>
+    /// <remarks>The wrapped <see cref="ActionRenderer"/> will not receive any
+    /// <see cref="IActionRenderer{T}.RenderActionError"/> and
+    /// <see cref="IActionRenderer{T}.UnrenderActionError"/> events except for block actions,
+    /// which do not belong to any transactions.
+    /// </remarks>
+    /// <typeparam name="T">An <see cref="IAction"/> type.  It should match to
+    /// <see cref="BlockChain{T}"/>'s type parameter.</typeparam>
+    public sealed class AtomicActionRenderer<T> : IActionRenderer<T>
+        where T : IAction, new()
+    {
+        private readonly List<(IAction, IActionContext, IAccountStateDelta)> _eventBuffer;
+        private TxId? _lastTxId;
+        private bool _errored;
+
+        /// <summary>
+        /// Creates a new <see cref="AtomicActionRenderer{T}"/> instance decorating the given
+        /// <paramref name="actionRenderer"/>.
+        /// </summary>
+        /// <param name="actionRenderer">The inner action renderer which has the <em>actual</em>
+        /// implementations and expects to receive no <see cref="RenderActionError"/>/<see
+        /// cref="UnrenderActionError"/> events.</param>
+        public AtomicActionRenderer(IActionRenderer<T> actionRenderer)
+        {
+            ActionRenderer = actionRenderer;
+            _lastTxId = null;
+            _eventBuffer = new List<(IAction, IActionContext, IAccountStateDelta)>();
+            _errored = false;
+        }
+
+        /// <summary>
+        /// The inner action renderer which has the <em>actual</em> implementations and expects to
+        /// receive no <see cref="RenderActionError"/>/<see cref="UnrenderActionError"/> events.
+        /// </summary>
+        public IActionRenderer<T> ActionRenderer { get; }
+
+        /// <inheritdoc cref="DelayedRenderer{T}.RenderBlock(Block{T}, Block{T})"/>
```

Comment:
Is there a reason to inherit document of `DelayedRenderer` not `IRenderer` for only this method, like other methods do? In my sight, it seems not related with `DelayedRenderer{T}.RenderBlock` method.

Reply:
It was a mistake made during I copy-and-pasted things… 🙄 Anyway amended: 69b6474d7914f3929154ed9a6474c5f8828f6661.